Cubical/Structures/Semigroup.agda

```diff
@@ -3,80 +3,168 @@ module Cubical.Structures.Semigroup where
 
 open import Cubical.Foundations.Prelude
 open import Cubical.Foundations.Equiv
+open import Cubical.Foundations.Equiv.HalfAdjoint
 open import Cubical.Foundations.HLevels
+open import Cubical.Foundations.Isomorphism
+open import Cubical.Foundations.Univalence
+open import Cubical.Foundations.Transport
+open import Cubical.Foundations.SIP renaming (SNS-PathP to SNS)
+
 open import Cubical.Data.Sigma
 
-open import Cubical.Foundations.SIP renaming (SNS-PathP to SNS)
 open import Cubical.Structures.Pointed
 open import Cubical.Structures.NAryOp
 
 private
   variable
     ℓ ℓ' : Level
 
-raw-semigroup-structure : Type ℓ → Type ℓ
-raw-semigroup-structure X = X → X → X
+-- Semygroup as a record, inspired by the standard library.
+-- Note that as we are using ≡ for all equations the IsMagma record would only contain isSet A if we had it.
+record IsSemigroup {A : Type ℓ} (_·_ : A → A → A) : Type ℓ where
 
-raw-semigroup-is-SNS : SNS {ℓ} raw-semigroup-structure _
-raw-semigroup-is-SNS = binaryFunSNS pointed-iso pointed-is-SNS
+  field
+    is-set : isSet A -- TODO: rename and move to IsMagma?
+    assoc  : (x y z : A) → x · (y · z) ≡ (x · y) · z
 
-semigroup-axioms : (X : Type ℓ) → raw-semigroup-structure X → Type ℓ
-semigroup-axioms X _·_ = isSet X ×
-                         ((x y z : X) → (x · (y · z)) ≡ ((x · y) · z))
+record Semigroup : Type (ℓ-suc ℓ) where
 
-semigroup-structure : Type ℓ → Type ℓ
-semigroup-structure = add-to-structure (raw-semigroup-structure) semigroup-axioms
+  -- TODO: add a constructor?
 
-Semigroup : Type (ℓ-suc ℓ)
-Semigroup {ℓ} = TypeWithStr ℓ semigroup-structure
+  infixl 7 _·_
 
--- Operations for extracting components
+  field
+    Carrier     : Type ℓ
+    _·_         : Carrier → Carrier → Carrier
+    isSemigroup : IsSemigroup _·_
+
+  open IsSemigroup isSemigroup public
+
+
+record SemigroupIso (M N : Semigroup {ℓ}) : Type ℓ where
+
+  -- How can this be written nicer without all the qualified names?
```

Comment:
Does anyone know the answer to this? @Saizan maybe?

Reply:
Untested, but you should be able to do something like this (adding those two lines between `where` and `field`)
```agda
open Semigroup M renaming (Carrier to [M]; _·_ to _·M_)
open Semigroup N renaming (Carrier to [N]; _·_ to _·N_)
```
and then simply use `[M]`, `[N]`, `_·M_` and `_·N_` (or whatever other names you choose).

Reply:
Something else that might work and doesn’t require renaming explicitly all the fields is (also untested)
```agda
module M = Semigroup M
module N = Semigroup N
```
and then use `M.Carrier`, `N.Carrier`, `x M.· y`, and so on.
Not entirely sure it’s ok to reuse the same names `M` and `N` for those modules, but it’s worth trying.

Reply:
the namespace for modules is separate, so you can reuse `M` and `N` for the module names.

Reply:
Thanks! I'll try both of these and figure out which I prefer

Reply:
Both worked fine, but I liked the `module M = Semigroup M` solution a bit more